Mathlib/LinearAlgebra/TensorProduct/LinearMap.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2023 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser, Christopher Hoskin
+-/
+import Mathlib.LinearAlgebra.Dual
+import Mathlib.LinearAlgebra.TensorProduct.Tower
+
+/-!
+# Bilinear maps on a Tensor Product
+
+## Main definitions
+
+* `LinearMap.tensorDistrib (B₁ ⊗ₜ B₂)`: the bilinear map on `M₁ ⊗ M₂` constructed by applying
+  `B₁` on `M₁` and `B₂` on `M₂`.
+-/
+
+suppress_compilation
+
+universe u v w uι uR uA uM₁ uM₂
+
+variable {ι : Type uι} {R : Type uR} {A : Type uA} {M₁ : Type uM₁} {M₂ : Type uM₂}
+
+open TensorProduct
+
+namespace LinearMap
+
+section CommSemiring
+
+variable [CommSemiring R] [CommSemiring A]
+variable [AddCommMonoid M₁] [AddCommMonoid M₂]
+variable [Algebra R A] [Module R M₁] [Module A M₁]
+variable [IsScalarTower R A M₁]
+variable [Module R M₂]
+
+variable (R A) in
+/-- The tensor product of two bilinear maps injects into bilinear maps on tensor products.
+
+Note this is heterobasic; the bilinear map on the left can take values in an (commutative) algebra
+over the ring in which the right bilinear map is valued. -/
+def tensorDistrib : (M₁ →ₗ[A] M₁ →ₗ[A] A) ⊗[R] (M₂ →ₗ[R] M₂ →ₗ[R] R) →ₗ[A]
+    ((M₁ ⊗[R] M₂) →ₗ[A] (M₁ ⊗[R] M₂) →ₗ[A] A) :=
+  ((TensorProduct.AlgebraTensorModule.tensorTensorTensorComm R A M₁ M₂ M₁ M₂).dualMap
+    ≪≫ₗ (TensorProduct.lift.equiv A (M₁ ⊗[R] M₂) (M₁ ⊗[R] M₂) A).symm).toLinearMap
+  ∘ₗ TensorProduct.AlgebraTensorModule.dualDistrib R _ _ _
+  ∘ₗ (TensorProduct.AlgebraTensorModule.congr
+    (TensorProduct.lift.equiv A M₁ M₁ A)
+    (TensorProduct.lift.equiv R _ _ _)).toLinearMap
+
+@[simp]
+theorem tensorDistrib_tmul (B₁ : M₁ →ₗ[A] M₁ →ₗ[A] A) (B₂ : M₂ →ₗ[R] M₂ →ₗ[R] R)
+    (m₁ : M₁) (m₂ : M₂) (m₁' : M₁) (m₂' : M₂) :
+    LinearMap.tensorDistrib R A (B₁ ⊗ₜ B₂) (m₁ ⊗ₜ m₂) (m₁' ⊗ₜ m₂')
+      = B₂ m₂ m₂' • B₁ m₁ m₁' :=
+  rfl
+
+/-- The tensor product of two bilinear forms, a shorthand for dot notation. -/
+@[reducible]
+protected def tmul (B₁ : M₁ →ₗ[A] M₁ →ₗ[A] A) (B₂ : M₂ →ₗ[R] M₂ →ₗ[R] R) :
+    (M₁ ⊗[R] M₂) →ₗ[A] (M₁ ⊗[R] M₂) →ₗ[A] A :=
+  LinearMap.tensorDistrib R A (B₁ ⊗ₜ[R] B₂)
+
+attribute [ext] TensorProduct.ext in
+/-- A tensor product of symmetric bilinear forms is symmetric. -/
+lemma IsSymm.tmul {B₁ : M₁ →ₗ[A] M₁ →ₗ[A] A} {B₂ : M₂ →ₗ[R] M₂ →ₗ[R] R}
+    (hB₁ : B₁.IsSymm) (hB₂ : B₂.IsSymm) : (B₁.tmul B₂).IsSymm := by
+  rw [LinearMap.isSymm_iff_eq_flip]
+  ext x₁ x₂ y₁ y₂
+  exact congr_arg₂ (HSMul.hSMul) (hB₂ x₂ y₂) (hB₁ x₁ y₁)
```

Comment:
Can you use this to prove the other `IsSymm.tmul`?

Reply:
I started looking at this. The BilinForm version of `IsSymm` doesn't require the semiring to be commutative. To do this properly we'd want the second argument to be a right module over `R` instead of a left module. Alternatively we could pursue the same strategy as we did with quadratic forms and remove the non-commutative semi-ring case for now.

#10422

Reply:
Turns out I was wrong - it is possible to prove `BilinForm.IsSymm.tmul` using `LinearMap.IsSymm.tmul` without changing the definition of `BilinForm.IsSymm`. I've pushed a commit to that effect.